src/box/txn_limbo.h

```diff
@@ -280,6 +280,16 @@ txn_limbo_is_replica_outdated(struct txn_limbo *limbo, uint32_t replica_id)
 	return res;
 }
 
+/** Return id of the node owning the synchro queue. */
+static inline uint32_t
+txn_limbo_owner_id(struct txn_limbo *limbo)
+{
+	latch_lock(&limbo->promote_latch);
+	uint32_t owner_id = limbo->owner_id;
+	latch_unlock(&limbo->promote_latch);
```

Comment:
Why do you need the lock to get the owner?

Reply:
I don't need it right now.
Actually `txn_limbo_owner_id` is only called in `applier_synchro_filter_tx`, which doesn't yield after `txn_limbo_is_replica_outdated`.
And `txn_limbo_is_replica_outdated` already takes the lock, so we are fine even without taking the lock again.

But I think this is less error-prone to make owner_id getter take the lock right now.
This way we don't care if there are some new yields added to `applier_synchro_filter_tx` later.

Now, forget `applier_synchro_filter_tx`. Why would `txn_limbo_owner_id` need a lock at all? For the same reason `txn_limbo_is_replica_outdated` needs it. To be sure that there are no ongoing promotions/demotions. Otherwise some tx might pass the owner_id check while there's already a new PROMOTE being written.

Reply:
Yeah, I see now, thanks for the explanation. I keep forgetting stuff. My worry here is that we just take lock/unlock too often, I am afraid about the perf. Also what if indeed there would be a yield between the is-outdated check and the owner id get? Could the replica during the yield become non-outdated? - I worry about atomicity here. Maybe we could just inline `txn_limbo_is_replica_outdated` call here and get the owner id in the same lock as we took for getting the term? I am fine with the current solution too, for now it works good I suppose.

Reply:
> Also what if indeed there would be a yield between the is-outdated check and the owner id get? Could the replica during the yield become non-outdated? - I worry about atomicity here. 

I think this would be possible, indeed.

> Maybe we could just inline txn_limbo_is_replica_outdated call here and get the owner id in the same lock as we took for getting the term?

This is a good option IMO. Take a look (spread over one new and one existing commit):
```diff
diff --git a/src/box/applier.cc b/src/box/applier.cc
index ccac9098a..c842e44a8 100644
--- a/src/box/applier.cc
+++ b/src/box/applier.cc
@@ -1182,13 +1182,19 @@ apply_final_join_tx(uint32_t replica_id, struct stailq *rows)
 static void
 applier_synchro_filter_tx(struct stailq *rows)
 {
+	latch_lock(&txn_limbo.promote_latch);
+	auto guard = make_scoped_guard([] {
+		latch_unlock(&txn_limbo.promote_latch);
+	});
 	struct xrow_header *row;
 	/*
 	 * It  may happen that we receive the instance's rows via some third
 	 * node, so cannot check for applier->instance_id here.
 	 */
 	row = &stailq_first_entry(rows, struct applier_tx_row, next)->row;
-	if (!txn_limbo_is_replica_outdated(&txn_limbo, row->replica_id))
+	uint64_t term = txn_limbo_replica_term(&txn_limbo, row->replica_id);
+	assert(term <= txn_limbo.promote_greatest_term);
+	if (term == txn_limbo.promote_greatest_term)
 		return;
 
 	/*
@@ -1207,7 +1213,7 @@ applier_synchro_filter_tx(struct stailq *rows)
 		&stailq_last_entry(rows, struct applier_tx_row, next)->row;
 	if (!last_row->wait_sync) {
 		if (iproto_type_is_dml(last_row->type) &&
-		    txn_limbo_owner_id(&txn_limbo) != REPLICA_ID_NIL) {
+		    txn_limbo.owner_id != REPLICA_ID_NIL) {
 			tnt_raise(ClientError, ER_SPLIT_BRAIN,
 				  "got an async transaction from an old term");
 		}
diff --git a/src/box/txn_limbo.h b/src/box/txn_limbo.h
index 99b7e5df8..5fb858295 100644
--- a/src/box/txn_limbo.h
+++ b/src/box/txn_limbo.h
@@ -266,30 +266,6 @@ txn_limbo_replica_term(const struct txn_limbo *limbo, uint32_t replica_id)
 	return vclock_get(&limbo->promote_term_map, replica_id);
 }
 
-/**
- * Check whether replica with id @a source_id is too old to apply synchronous
- * data from it.
- */
-static inline bool
-txn_limbo_is_replica_outdated(struct txn_limbo *limbo, uint32_t replica_id)
-{
-	latch_lock(&limbo->promote_latch);
-	uint64_t v = vclock_get(&limbo->promote_term_map, replica_id);
-	bool res = v < limbo->promote_greatest_term;
-	latch_unlock(&limbo->promote_latch);
-	return res;
-}
-
-/** Return id of the node owning the synchro queue. */
-static inline uint32_t
-txn_limbo_owner_id(struct txn_limbo *limbo)
-{
-	latch_lock(&limbo->promote_latch);
-	uint32_t owner_id = limbo->owner_id;
-	latch_unlock(&limbo->promote_latch);
-	return owner_id;
-}
-
 /**
  * Return the last synchronous transaction in the limbo or NULL when it is
  * empty.

```